File: ruby/bird-count/bird_count.rb
# frozen_string_literal: true

class BirdCount
  LAST_WEEK_BIRDS = [0, 2, 5, 3, 7, 8, 4].freeze
  def self.last_week
    LAST_WEEK_BIRDS
  end

  def initialize(birds_per_day)
    @birds_per_day = birds_per_day
  end

  def yesterday
    @birds_per_day[-2]
  end

  def total
    @birds_per_day.sum
  end

  def busy_days
    @birds_per_day.count { |x| x >= 5 }
  end

  def day_without_birds?
    @birds_per_day.any?(&:zero?)
  end
end

File: elixir/bird-count/lib/bird_count.ex
defmodule BirdCount do
  def today([]) do
    nil
  end

  def today([today | _tail]) do
    today
  end

  def increment_day_count([]) do
    [1]
  end

  def increment_day_count([today | tail]) do
    [today + 1 | tail]
  end

  def has_day_without_birds?(list) do
    0 in list
  end

  def total([]) do
    0
  end

  def total([today | tail]) do
    today + total(tail)
  end

  def busy_days([]) do
    0
  end

  def busy_days([today | tail]) do
    cond do
      today >= 5 -> 1 + busy_days(tail)
      true -> busy_days(tail)
    end
  end
end
